src/misc/degree_sequence.cpp

```diff
@@ -68,123 +70,282 @@ template<typename T> inline bool degree_less(const T &a, const T &b) {
+ * [0][1][2][3][5]          4-2
+ *        0                 3-0
+ *        2
+ *        1
+ *
+ * The process continues until we are unable to find either a non-zero "hub" node
+ * (algorithm complete and sequence is graphical) or enough non-zero "spoke" nodes once a
+ * hub has been selected (sequence is non-graphical).
+ */
+static igraph_error_t igraph_i_havel_hakimi(const igraph_vector_int_t *degseq,
+                                            igraph_vector_int_t *edges,
+                                            igraph_realize_degseq_t method) {
+    igraph_integer_t n_nodes = igraph_vector_int_size(degseq);
 
-        vd_pair vd = *pt;
-        vertices.erase(pt);
+    // ----- upfront error/graphicality checks ----- //
+    if (n_nodes == 0 || (n_nodes == 1 && VECTOR(*degseq)[0] == 0)) {
+        return IGRAPH_SUCCESS;
+    }
 
-        if (vd.degree == 0) {
-            continue;
+    for (igraph_integer_t i = 0; i < n_nodes; i++) {
+        igraph_integer_t deg = VECTOR(*degseq)[i];
+        if (deg >= n_nodes) {
+            IGRAPH_ERROR("The given degree sequence cannot be realized as a simple graph.", IGRAPH_EINVAL);
         }
+    }
 
-        igraph_integer_t k;
-        vlist::iterator it;
-        for (it = vertices.begin(), k = 0;
-             k != vd.degree && it != vertices.end();
-             ++it, ++k) {
-            if (--(it->degree) < 0) {
-                goto fail;
-            }
-
-            VECTOR(*edges)[2 * (ec + k)] = vd.vertex;
-            VECTOR(*edges)[2 * (ec + k) + 1] = it->vertex;
+    // ----- main Havel-Hakimi loop ----- //
+    // O(V + alpha(V) * E)
+    // O(V + E) for the LARGEST_FIRST method
+    igraph_vector_int_t seq;
+    IGRAPH_CHECK(igraph_vector_int_init_copy(&seq, degseq));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &seq);
+
+    HavelHakimiList vault(&seq);
+
+    igraph_integer_t n_edges_added = 0;
+    for (igraph_integer_t i = 0; i < n_nodes; i++) {
+        // hub node selection
+        vd_pair hub;
+        if (method == IGRAPH_REALIZE_DEGSEQ_SMALLEST) {
+            if (!vault.get_min_node(/* out param */hub)) break;
+            vault.remove_min_node();
         }
-        if (it == vertices.end() && k < vd.degree) {
-            goto fail;
+        else if (method == IGRAPH_REALIZE_DEGSEQ_LARGEST) {
+            if (!vault.get_max_node(/* out param */hub)) break;
+            vault.remove_max_node();
         }
-
-        ec += vd.degree;
+        else if (method == IGRAPH_REALIZE_DEGSEQ_INDEX) {
+            igraph_integer_t degree = VECTOR(seq)[i];
+            hub = vd_pair{i, degree};
+            vault.buckets[degree].count--;
+        }
+        VECTOR(seq)[hub.vertex] = 0;
+
+        // spoke nodes selection
+        igraph_vector_int_t spokes;
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&spokes, 0);
+        IGRAPH_CHECK(igraph_vector_int_reserve(&spokes, hub.degree));
+        IGRAPH_CHECK(vault.get_spokes(hub.degree, seq, spokes));
```

Comment:
Let's refactor this:

 - `get_spokes()` should not rely on `spokes` being an empty vector. It should use `vector_int_clear()` to make sure it's empty before pushing to it.
 - The storage should be reserved by `get_spokes()`, as that's where the relevant logic is located.

The `spokes` vector is repeatedly allocated and destroyed in a loop. This is inefficient. Keep in mind: a memory allocation is always slow, no matter how little memory we are allocating!

Solution: Allocate `spokes` once outside the loop, and re-use it multiple times.